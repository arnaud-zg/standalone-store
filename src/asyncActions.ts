import { AnyAction, createStore } from 'redux'
import { ActionCreator, isActionOf } from 'typesafe-actions'
import { StandaloneStore } from '.'

interface IResolve<TState> {
  action: AnyAction
  state: TState
}

export const dispatchActionsAndWaitResponse = <TState, TSelectorResponse>({
  actionsDispatch,
  actionCreatorsResolve,
  store,
  selector,
}: {
  store: ReturnType<typeof createStore>
  actionsDispatch: AnyAction[]
  actionCreatorsResolve: ActionCreator[]
  selector: (state: TState) => TSelectorResponse
}) => {
  const standaloneStore = new StandaloneStore<TState>({ store })

  return !actionsDispatch.length
    ? Promise.reject('You should at least give one action.')
    : new Promise((resolve: ({ action, state }: IResolve<TState>) => void) => {
        actionCreatorsResolve.forEach(actionResolve => {
          standaloneStore.subscribe((action, state) => {
            if (isActionOf(actionResolve, action)) {
              resolve({ action, state })
            }
          })
        })
        actionsDispatch.forEach(action => {
          standaloneStore.dispatchAction(action)
        })
      }).then(({ state }: { state: TState }) => {
        standaloneStore.listenersClear()
        return selector(state)
      })
}
