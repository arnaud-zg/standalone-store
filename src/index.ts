export * from './asyncActions'
export * from './standaloneStore'
