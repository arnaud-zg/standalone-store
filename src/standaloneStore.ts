import { AnyAction, StoreCreator } from 'redux'

type Listener<TStoreState> = (
  action: AnyAction | null,
  state: TStoreState
) => void

export class StandaloneStore<TStoreState> {
  private listeners: Array<Listener<TStoreState>>
  private store: ReturnType<StoreCreator>
  private lastAction: AnyAction | null

  constructor({ store }: { store: ReturnType<StoreCreator> }) {
    this.lastAction = null
    this.listeners = []
    this.store = store

    this.store.subscribe(this.storeUpdate)
  }

  dispatchAction = (action: AnyAction) => {
    this.lastAction = action
    this.store.dispatch(action)
  }

  storeUpdate = () => {
    const state = this.store.getState() as TStoreState

    if (this.listeners.length) {
      this.listeners.forEach(listener => {
        listener(this.lastAction, state)
      })
    }
  }

  subscribe(listener: Listener<TStoreState>) {
    this.listeners.push(listener)
  }
}
