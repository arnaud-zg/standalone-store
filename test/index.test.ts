import * as projectPackage from '../src'

describe('projectPackage', () => {
  it('should have a list of method', () => {
    expect(projectPackage).toMatchSnapshot()
  })
})
