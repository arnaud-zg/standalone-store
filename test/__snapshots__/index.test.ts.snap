// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectPackage should have a list of method 1`] = `
Object {
  "StandaloneStore": [Function],
}
`;
